ilastik/applets/neuralNetwork/modelStateControl.py

```diff
@@ -0,0 +1,414 @@
+        self._model_source = model_source
+
+    def dropEvent(self, dropEvent):
+        urls = dropEvent.mimeData().urls()
+        self.clear()
+        self.setPlainText(urls[0].toLocalFile())
+
+    def dragEnterEvent(self, event):
+        # Only accept drag-and-drop events that consist of a single local file.
+        if not event.mimeData().hasUrls():
+            return
+        urls = event.mimeData().urls()
+        if len(urls) == 1 and all(url.isLocalFile() for url in urls):
+            event.acceptProposedAction()
+
+    def clear(self):
+        super().clear()
+        self.setEnabled(True)
+        self.setTextInteractionFlags(Qt.TextEditorInteraction)
+        self.setToolTip(
+            "Copy/Paste a bioimage.io model doi or nickname, or\ndrag-and-drop a downloaded bioimage.io model.zip file."
+        )
+        self._model_source = None
+        self._initialized = False
+
+    def askClear(self):
+        if (
+            QMessageBox.question(self, "Remove model", "Do you want to remove the model from the project file?")
+            == QMessageBox.Yes
+        ):
+            self.clear()
+            self.modelDeleted.emit()
+
+    def setModelIncompatibleState(self, model_source, model_info, error_message):
+        self.setToolTip(f"Model not compatible with data:\n\n{error_message}")
+        self.setModelInfo(model_source, model_info, template=Template(f"<red>{display_template}</red>"))
+
+    def setEmptyState(self):
+        self.clear()
+
+    def setModelDataAvailableState(self, model_source, model_info):
+        self.btn_container.setEnabled(True)
+        self.setTextInteractionFlags(Qt.NoTextInteraction)
+        self.setToolTip("Remove the model by clicking the 'x' in the upper right corner.")
+        self.setModelInfo(model_source, model_info)
+
+    def setReadyState(self, model_source, model_info):
+        self.btn_container.setEnabled(False)
+        self.setToolTip("Stop the model to make changes.")
+        self.setModelInfo(model_source, model_info)
+
+
+class ModelStateControl(QWidget):
+    uploadDone = pyqtSignal()
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        self.threadRouter = ThreadRouter(self)
+        self._preDownloadChecks = set()
+
+        layout = QVBoxLayout()
+
+        self.modelSourceEdit = ModelSourceEdit(self)
+        self.statusLabel = QLabel(self)
+        self.modelControlButton = QToolButton(self)
+        self.modelControlButton.setText("...")
+        self.modelControlButton.setToolTip("Click here to check model details, initialize, or un-initialize the model")
+
+        bottom_layout = QHBoxLayout()
+        bottom_layout.addWidget(self.statusLabel)
+        bottom_layout.addStretch()
+        bottom_layout.addWidget(self.modelControlButton)
+        layout.addWidget(self.modelSourceEdit)
+        layout.addLayout(bottom_layout)
+        self.setLayout(layout)
+
+    def setTiktorchController(self, tiktorchController):
```

Comment:
Maybe use a setter property here?

Reply:
Hm, we're still on qt5.12 - there was a bug that was related to having properties that would be accessed on instantiation. While one could definitely use them safely, I still prefer not to use them at all...

xref: https://github.com/ilastik/volumina/pull/261